cub/device/device_memcpy.cuh

```diff
@@ -0,0 +1,158 @@
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions are met:
+ *     * Redistributions of source code must retain the above copyright
+ *       notice, this list of conditions and the following disclaimer.
+ *     * Redistributions in binary form must reproduce the above copyright
+ *       notice, this list of conditions and the following disclaimer in the
+ *       documentation and/or other materials provided with the distribution.
+ *     * Neither the name of the NVIDIA CORPORATION nor the
+ *       names of its contributors may be used to endorse or promote products
+ *       derived from this software without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND
+ * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
+ * WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+ * DISCLAIMED. IN NO EVENT SHALL NVIDIA CORPORATION BE LIABLE FOR ANY
+ * DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
+ * (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+ * LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND
+ * ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+ * SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ *
+ ******************************************************************************/
+
+/**
+ * \file
+ * cub::DeviceMemcpy provides device-wide, parallel operations for copying data.
+ */
+
+#pragma once
+
+#include <cub/config.cuh>
+#include <cub/device/dispatch/dispatch_batch_memcpy.cuh>
+
+#include <cstdint>
+#include <type_traits>
+
+CUB_NAMESPACE_BEGIN
+
+/**
+ * @brief cub::DeviceMemcpy provides device-wide, parallel operations for copying data.
+ * \ingroup SingleModule
+ */
+struct DeviceMemcpy
+{
+  /**
+   * @brief Copies data from a batch of given source buffers to their corresponding destination buffer.
+   * @note If any input buffer aliases memory from any output buffer the behavior is undefined. If any output buffer
+   * aliases memory of another output buffer the behavior is undefined. Input buffers can alias one another.
+   *
+   * @par Snippet
+   * The code snippet below illustrates usage of DeviceMemcpy::Batch for mutating strings withing a single string
+   * buffer.
+   * @par
+   * @code
+   *  struct GetPtrToStringItem
+   * {
+   *   __host__ __device__ __forceinline__ void *operator()(uint32_t index)
+   *   {
+   *     return &d_string_data_in[d_string_offsets[index]];
+   *   }
+   *   char *d_string_data_in;
+   *   uint32_t *d_string_offsets;
+   * };
+   *
+   * struct GetStringItemSize
+   * {
+   *   __host__ __device__ __forceinline__ uint32_t operator()(uint32_t index)
+   *   {
```

Comment:
For some reason this example only works when I compile with `-G -g`. Otherwise I get strange output. Could you check if there's some kind of race in the algorithm or a compiler issue?

Reply:
There was a logical issue with pre-Volta architectures / tuning policy in the `BitPackedCounter`. This should be fixed now.